### PR TITLE
Login with Amazon

### PR DESCRIPTION
**Pull request by dmarvp**
relates #518 
- added the sign in with custom token firebase method
- added the calls to virtual device and source api related to the amazon login
- when logging in, an access token is used to get the amazon profile
- we call source-api with that profile to get an auth token
- if no user is present, one is created
- after logging in with the auth token, a dispatch to setAmazonFlow is made

### CODE REVIEW

File: harness/index.html [hunk 1]
```diff
@@ -69,6 +69,19 @@
 <script src="https://unpkg.com/react-dom@15.3.2/dist/react-dom.js"></script>
 <!-- Import the dashboard dependency -->
 <script src="/scripts/dashboard.js"></script>
+<!-- Login with Amazon -->
+<div id="amazon-root"></div>
+<script>
+    window.onAmazonLoginReady = function() {
+    amazon.Login.setClientId('amzn1.application-oa2-client.675b266c8bca4d408e5dceed285e18bd');
```
Comment (jperata) [inline harness/index.html]: use an env vairable

File: src/actions/session.ts [hunk 2]
```diff
@@ -16,13 +16,25 @@ export type SetUser = {
   user: User | undefined
 };
 
+export type SetAmazonFlow = {
```
Comment (jperata) [inline src/actions/session.ts]: AmazonFlowFlag to differentiate from the fucntion

File: src/pages/LoginPage.tsx [hunk 3]
```diff
@@ -93,6 +97,9 @@ export class LoginPage extends React.Component<LoginPageProps, LoginPageState> {
 
     handleFormLoginWithAmazon() {
         this.props.loginWithAmazon()
+            .then((user) => {
```
Comment (jperata) [inline src/pages/LoginPage.tsx]: try {        
  await this.props.loginWithAmazon();
  this.props.setAmazonFlow(true); 
} catch( err: Error) {
                 this.state.error = err.message;
                 this.setState(this.state);
}

File: src/services/auth.ts [hunk 4]
```diff
@@ -19,8 +20,31 @@ namespace auth {
...
+        const accessToken = await amazonAuthorize(options);
+        const amazonProfile: any = await amazonRetrieveProfile(accessToken);
+        const authResponse = await source.getAuthToken(amazonProfile.PrimaryEmail, amazonProfile.Name);
+        return loginWithCustomToken(authResponse.authToken);
+    }
+
+    function amazonAuthorize(options: any): Promise<string> {
+        return new Promise(function (resolve, reject) {
+            globalWindow.amazon.Login.authorize(options, function (result: any) {
+                if (result.error) { reject(result.error); };
```
Comment (jperata) [inline src/services/auth.ts]: while reject here ensure state of the promise is always reject, it doesn't actually end the code flow, 
```
 if (result.error) {
 reject(result.error); 
 return;
}; 
```
will have the same effect but any addition after the return is ensured to not be run

File: src/services/auth.ts [hunk 5]
```diff
@@ -19,8 +20,31 @@ namespace auth {
...
+                resolve(result.access_token);
+            });
+        });
+    }
+
+    function amazonRetrieveProfile(accessToken: string) {
+        return new Promise(function (resolve, reject) {
+            globalWindow.amazon.Login.retrieveProfile(accessToken, function (result: any) {
+                if (result.error) { reject(result.error); };
```
Comment (jperata) [inline src/services/auth.ts]: same

File: src/services/auth.ts [hunk 6]
```diff
@@ -135,6 +159,20 @@ namespace auth {
             });
     }
 
+    export function loginWithCustomToken(token: string, auth: remoteservice.auth.Auth = remoteservice.defaultService().auth(), localStorage: LocalStorage = new BrowserStorage()): Promise<User> {
+        return auth.signInWithCustomToken(token)
```
Comment (jperata) [inline src/services/auth.ts]:
```
const user: remoteservice.user.User = await auth.signInWithCustomToken(token);
ReactGA.event({
   category: "Authorization",
   action: "Login With Custom Token"
});
let modelUser: User = new FirebaseUser(user);
identify(modelUser, "custom token");
ocalStorage.setItem("user", JSON.stringify(modelUser));
return modelUser;
```

File: src/services/auth.ts [hunk 7]
```diff
@@ -135,6 +159,20 @@ namespace auth {
...
+            .then(function (user: remoteservice.user.User) {
+                ReactGA.event({
+                    category: "Authorization",
+                    action: "Login With Custom Token"
+                });
+                let modelUser: User = new FirebaseUser(user);
```
Comment (jperata) [inline src/services/auth.ts]: const if identify doesn't change modelUser

File: src/services/source.ts [hunk 8]
```diff
@@ -73,6 +80,33 @@ export namespace source {
...
     }
 
+    /**
+     * This service will return an authorization token for a given email.  If the user does not exist,
+     * it will create it first.
+     *
+     * @param email the user's email.
+     * @param displayName the user's display name (optional)
+     */
+    export function getAuthToken(email: string, displayName: string): Promise<AuthResult> {
```
Comment (jperata) [inline src/services/source.ts]: shouldnt this be diplayName?: string
according to your comment above

File: src/services/source.ts [hunk 9]
```diff
@@ -73,6 +80,33 @@ export namespace source {
...
+        const query: Query = new Query();
+        query.add({parameter: "email", value: email});
+        query.add({parameter: "displayName", value: displayName});
+
+        return fetch(AUTH_TOKEN_URL, {
```
Comment (jperata) [inline src/services/source.ts]: const result = await fetch ...

File: harness/index.html (hunk 1 above)
Comment (dmarvp) [reply harness/index.html]: Can't use an env variable in plain html. We should be safe, as the LWA uses both the client id and a whitelist to get the urls that can use that client id.

File: src/services/auth.ts [hunk 10]
```diff
@@ -4,7 +4,7 @@ import { FirebaseUser } from "../models/user";
 import { User, UserDetails } from "../models/user";
 import { BrowserStorage, LocalStorage } from "../store/local-storage";
 import browser from "../utils/browser";
-import { remoteservice } from "./remote-service";
+import remoteService, { remoteservice } from "./remote-service";
```
Comment (jperata) [inline src/services/auth.ts]: remoteService (S capital) is just an alias, using remoteservice everywhere should work

**Pull request merged**